### ISSUE TITLE
Adding filter & action hooks

### ISSUE THREAD

**Issue by IanDelMar** [understrap/understrap#887]
# Suggestion: Adding filter & action hooks
I really like UnderStrap and appreciate all the work that has been put into it so far. However, I often end up adding or replacing hard-coded classes in my child themes because I'm to lazy to do it via scss/css and I think that's what all the pre-made bootstrap classes actually are for. 
For example, if you want to have `.navbar-light` instead of `.navbar-dark`, you could either rewrite the header.php or swap the SASS variables `$dark` and `$light`. The first means that the file will not be subject to updates, while the latter makes the naming pretty counter-intuitive. Another example, if you want to add a top bar then again you are forced to rewrite the header.php.
Both situations could be avoided if there were filter and action hooks throughout the template files.

## Filter hooks for HTML classes of elements
For the filter hooks there could be something like this which dynamically adds filters
```php
/**
 * Display HTML classes for an element
 *
 * @param string $element The element for which to display the classes.
 */
function understrap_element_classes( $element ) {
	echo implode( ' ', understrap_get_element_classes( $element ) );
}

/**
 * Retrieve HTML classes for an element
 *
 * @param string $element The element for which to retrieve the classes.
 * @return array Maybe filtered array of classes.
 */
function understrap_get_element_classes( $element ) {
	$classes = understrap_get_default_classes( $element );
	$classes = apply_filters( "understrap_{$element}_classes", $classes );

	return array_map( 'esc_attr', $classes );
}

/**
 * Retrieve default HTML classes for an element
 *
 * @param string $element The element for which to retrieve the classes.
 * @return array $classes Maybe filtered array of classes.
 */
function understrap_get_default_classes( $element ) {
	$defaults = array(
		'container'      => 'container',
		'navbar'         => 'navbar-expand-md navbar-dark bg-primary',
		'sidebar-left'   => 'col-md-4',
		'sidebar-right'  => 'col-md-4',
		'sidebars-left'  => 'col-md-3',
		'sidebars-right' => 'col-md-3',
	);
	$defaults = $defaults[ $element ];
	$defaults = apply_filters( "understrap_{$element}_default_classes", $defaults );

	$classes = array();
	if ( $defaults ) {
		$classes = preg_split( '#\s+#', $defaults );
	}

	return $classes;
}
```
and in the header.php
```php
<nav class="navbar <?php understrap_element_classes( 'navbar' ); ?>">
```
which then can be used like so
```php
function child_navbar_light( $defaults ) {
	return 'navbar-expand-md navbar-light bg-primary';
}
add_filter( 'understrap_navbar_default_classes', 'child_navbar_light' );
```
giving 
```html
<nav class="navbar navbar-expand-md navbar-light bg-primary">
```
instead of 
```html
<nav class="navbar navbar-expand-md navbar-dark bg-primary">
```
or used to add classes like so
```php
function child_navbar_new_class( $classes ) {
	$classes[] = 'new-class';
	return $classes;
}
add_filter( 'understrap_navbar_classes', 'child_navbar_new_class' );
```
giving
```html
<nav class="navbar navbar-expand-md navbar-dark bg-primary new-class">
```
## Action hooks for adding elements
For the most flexible header.php action hooks could look like this
```php
<?php
/**
 * The header for our theme.
 *
 * Displays all of the <head> section and everything up till <div id="content">
 *
 * @package understrap
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}
?>
<!DOCTYPE html>
<html <?php language_attributes(); ?>>
<head>
	<meta charset="<?php bloginfo( 'charset' ); ?>">
	<?php do_action( 'understrap_head' ); ?>
	<?php wp_head(); ?>
</head>

<body <?php body_class(); ?>>

<?php do_action( 'understrap_before_page' ); ?>

<div class="site" id="page">

	<?php 
	do_action( 'understrap_before_header' );
	do_action( 'understrap_header' );
	do_action( 'understrap_after_header' );
	?>
```
with functions adding all the stuff like so
```php
/**
 * Add viewport meta to head
 */
function understrap_add_viewport() {
	?>
	<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
	<?php
}
add_action( 'understrap_head', 'understrap_add_viewport' );

/**
 * Add profile link to head
 */
function understrap_add_profile() {
	?>
	<link rel="profile" href="http://gmpg.org/xfn/11">
	<?php
}
add_action( 'understrap_head', 'understrap_add_profile' );

/**
 * Display header
 *
 * @return void
 */
function understrap_add_header() {
	?>
	<div id="wrapper-navbar" itemscope itemtype="http://schema.org/WebSite">
		<?php
		do_action( 'understrap_before_navbar' );
		do_action( 'understrap_navbar' );
		do_action( 'understrap_after_navbar' );
		?>
	</div>
	<?php
}
add_action( 'understrap_header', 'understrap_add_header' );

/**
 * Add skip to content link
 */
function understrap_add_skip_to_content() {
	?>
	<a class="skip-link sr-only sr-only-focusable" href="#content"><?php esc_html_e( 'Skip to content', 'understrap' ); ?></a>
	<?php
}
add_action( 'understrap_before_navbar', 'understrap_add_skip_to_content' );

/**
 * Display navbar
 */
function understrap_add_navbar() {
	?>
	<nav id="primary-navbar" class="navbar <?php understrap_element_classes( 'navbar' ); ?>">
		<?php
		$contained = 'container' === get_theme_mod( 'understrap_container_type' );
		if ( $contained ) {
			echo '<div class="container">';
		}
		do_action( 'understrap_navbar_elements' );
		if ( $contained ) {
			echo '</div>';
		}
		?>
	</nav>
	<?php
}
add_action( 'understrap_navbar', 'understrap_add_navbar' );

/**
 * Display brand (custom logo or site title)
 *
 * @return void
 */
function understrap_add_brand() {
	if ( ! has_custom_logo() ) { ?>

		<?php if ( is_front_page() && is_home() ) : ?>

			<h1 class="navbar-brand mb-0"><a rel="home" href="<?php echo esc_url( home_url( '/' ) ); ?>" title="<?php echo esc_attr( get_bloginfo( 'name', 'display' ) ); ?>" itemprop="url"><?php bloginfo( 'name' ); ?></a></h1>

		<?php else : ?>

			<a class="navbar-brand" rel="home" href="<?php echo esc_url( home_url( '/' ) ); ?>" title="<?php echo esc_attr( get_bloginfo( 'name', 'display' ) ); ?>" itemprop="url"><?php bloginfo( 'name' ); ?></a>

		<?php endif; ?>


	<?php } else {
		the_custom_logo();
	} 
}
add_action( 'understrap_navbar_elements', 'understrap_add_brand' );

/**
 * Add navbar toggler
 */
function understrap_add_navbar_toggler() {
	?>
	<button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNavDropdown" aria-controls="navbarNavDropdown" aria-expanded="false" aria-label="<?php esc_attr_e( 'Toggle navigation', 'understrap' ); ?>">
		<span class="navbar-toggler-icon"></span>
	</button>
	<?php
}
add_action( 'understrap_navbar_elements', 'understrap_add_navbar_toggler' );

/**
 * Display navigation menu
 */
function understrap_add_menu() {
	wp_nav_menu(
		array(
			'theme_location'  => 'primary',
			'container_class' => 'collapse navbar-collapse',
			'container_id'    => 'navbarNavDropdown',
			'menu_class'      => 'navbar-nav ml-auto',
			'fallback_cb'     => '',
			'menu_id'         => 'main-menu',
			'depth'           => 2,
			'walker'          => new Understrap_WP_Bootstrap_Navwalker(),
		)
	);
}
add_action( 'understrap_navbar_elements', 'understrap_add_menu' );
```

The header.php is only an example, the whole thing can and should be done with all the files (where appropriate). I think this would be a huge step forward in terms of flexibility.

# Question
What do you think? Before going through all the files and making a PR I want to be sure that it's not going to be downvoted or ignored.

**Comment by 0dp** [comment]
Only if you have to promise to write documentation too 😄

**Comment by IanDelMar** [comment]
@axlright 
> it always gets customized or re-written or re-ordered

That's exactly the point. I would prefer to do the customization via functions instead of changing the template files (a "Genesis like" approach). Wouldn't it be handy to re-order things just by hooking a function on to another action or changing its priority? Actually, that would be much more in line with the idea of a framework - which UnderStrap claimes to be.

> Adding in variables to the parent theme to anticipate some customization seems to me to make it more complicated when those customizations aren't sufficient and the child theme then needs to be edited further.

What are you thinking of?

> So maybe this puts me out of the conversation.

All comments are welcome and appreciated.

**Comment by pattonwebz** [comment]
The idea of action hooks placed strategically throughout the code like this works well for other themes but it can very quickly become cumbersome to add it to existing themes and get full coverage. It's also a very different way of developing themes for WordPress than the standard

Every value run through a filter at use time, every area of the theme tagged with an action hook etc. 

Going through the theme to add all of those calls individually, moving the code from the template file into a function and hooking it in then handling ordering via priority.  In some places it will involve implimention of several hooks per section for ordering - eg: `before_nav_open`, `after_nav_open`, `do_nav`, `before_nav_close`, `after_nav_close` and probably several more throughout the nav output (another exalted is post section has dozens of possible locations like before and after - as well as the actual action for in between them - for title, entry meta, post content, author box. 

To modify something inside any action hooks added that have existing content you still need to unhook and override existing functions... It's the same result as overriding a file but it's actually many more steps doing it via an action hook override than via a template override.

For reference Genesis does this for its themes but it does it to extremes and overrides the entire core templating system to make it happen. Overrides of the core template system would not be appropriate for .org directory themes as they are required to use core features where they exist.

**Comment by IanDelMar** [comment]
>The idea of action hooks placed strategically throughout the code like this works well for other themes but it can very quickly become cumbersome to add it to existing themes and get full coverage.

I totally agree and I'm not as naive as to think that this will be done in a single round rewrite. Most likely time after time there will pop up something what has not been considered in the first loop of implementation.

> It's the same result as overriding a file but it's actually many more steps doing it via an action hook override than via a template override.

Of course, it gives the same result. But I disagree that it's many more steps. Let's take the `header.php` as example again and assume that you want to reverse the order of the brand and the menu. You copy the file into your child theme folder, copy the brand code and paste it after the `wp_nav_menu()` and finally you adjust the menu classes. Given that the wp_nav_menu arguments are filterable then its two lines of code to re-order things (unhook function, re-hook function with different priority) and another five lines of code to filter the arguments. An example which makes it more obvious: Consider you want to display the categories in the current `.entry-meta` container instead of in the `.entry-footer`. You either have to rewrite the `content-single.php` putting the code into the file directly or you have to rewrite the `understrap_entry_footer()` and the `understrap_posted_on()` versus something like
```php
remove_action( 'understrap_entry_footer_elements', 'understrap_posted_in' );
add_action( 'understrap_entry_meta_elements', 'understrap_posted_in' );
```
I think the function approach is also more convenient if you want to re-use some things in other projects. 

Yes, Genesis is an extreme example and I should not have mentioned it as I don't suggest going that far. And yes, this could be an entry barrier for new users/developers, but this argument also holds the other way round. However, maybe this is really just a matter of taste and I'm not going to stubbornly insist that it's the better way for customizing a parent theme just because I'm biased towards my taste (especially as there is always the fallback to just create a another parent theme). Thank you for your thoughts on this issue!

**Comment by pattonwebz** [comment]
> Of course, it gives the same result. But I disagree that it's many more steps.
I'll concede that my original statement maybe was biased towards more complex scenarios and that your example does show some very good and simple use cases. I'll also say that I am not against filter based modification approaches (and that indeed they are sometimes a much preferable solution for certain kinds of changes).

> I totally agree and I'm not as naive as to think that this will be done in a single round rewrite.
If you are able/willing to support such iterative rewrites then I am not going to stand in the way of that. I don't really make decisions for this repo - but I like to poke my head into issues and throw some spanners in the works from time to time lol - however I do know from working with other themes that these methods can work very well when done thoughtfully :)

I am still mindful that this would be a significant rewrite and probably require a lot of someones time and effort to achieve. It would also be a pretty substantial change to the codebase for anyone else that may already be using the code and/or developing sites with it. < Just something to think about.

**Comment by axlright** [comment]
>Let's take the header.php as example again and assume that you want to reverse the order of the brand and the menu.
@IanDelMar My thing is that we're often totally re-writing the `header.php` for added navs, side navs that are in a totally different div than the collapse toggle, adding persistent callouts, adding CriticalCSS conditionals, etc. There are few projects that use the stock `header.php` even in a different order. So to me, I'd rather have free reign to re-write the markup and php than change variables. Again, that's just me.

**Comment by bbunlock** [comment]
I am relatively new to working with wordpress in regards to themes, I have made a relatively complex plugin though and I must admit I like the idea of using hooks and filters in the themes as I am used to using them in coding a plugin.

I also like the idea of the page templates as well so that being said I cannot say which I preffer other than to say I like both.

Kind regards

Wayne

**Comment by Thomas-A-Reinert** [comment]
Seems to totally make sense to me. But really, we´ll need to investigate and to weigh pro´s and con´s. 
Honestly, I could weigh in for both sides.

Anyway. Wanted to push that topic again due to housekeeping/cleaning up old issues and assigned some of the core-contributors. Hope you feel alright with that, just holla if you want to be excluded.

**Comment by pattonwebz** [comment]
Totally fine by me to be added to this issue :)

I do support the idea this issue is opened for - namely moving to a more actions/filters based approach. 

I still have the concerns I stated previously: 
1. The change is significant enough to hinder some existing users - a major version bump probably.
2. That it should be done in pieces so that any unexpected effects of swapping can be spotted.
3. That it will require a lot of time to do and to test.

I propose that we pick 1 or 2 template files (maybe header and footer) and rework those first to test how it could for into the understrap theme.

**Comment by axlright** [comment]
Happy to be added to this issue as well!

I would not use or benefit from moving to a more actions/filters based approach. My team ends up customizing everything. We don't ever use any of the built in page templates or sidebars. Everything gets built from scratch. We're using Understrap for Bootstrap and Gulp. Even the `$container` variable is a bit of a joke - for some reason we keep the variable but have never once changed the `understrap_container_type` customization.

I'd like to see more collaboration on the Gulp side - UnCSS, CriticalCSS - as well as responsive images and lazy loading and then Gutenberg blocks.

But maybe I'm not the intended user. I'm not looking to download the theme, customize it with the customizer, use stock page templates and publish a site.

So for me, I'd support the decision if the intended user is someone who would benefit from it. I would not use or benefit from it.

**Comment by pattonwebz** [comment]
@axlright this kind of change I feel would benefit your team very much and actually input from yourself throughout the planning process would probably help b is get a better result first time round from knowing more use cases.

I'm not at my computer anymore tonight but tomorrow I would like to have a longer discussion about this to see how best we can plan accordingly and to share Israel hour to do it.

**Comment by axlright** [comment]
@pattonwebz I'm always down for a conversation!

**Comment by stef-k** [comment]
@axlright judging from the issues you are participating, I think you and your team would benefit from an "`eject`" like solution such the one used in Angular - React. 

For teams wanting a barebones WP/Bootstrap starting package the script could be used to strip all the bells and whistles and leave only the basic. So instead of using time and energy to change existing functionality which may or may not impact existing workflows and users, someone could develop the "`eject`" functionality...

**Comment by adkhan11** [comment]
Lazy Loading Available now ?

**Comment by danemorgan** [comment]
I could go either way here. I appreciate aspects of both methods. I've used genesis extensively and I've used _s and Them,e Hybrid extensively. They each have different approaches that lend to different solutions.

As noted I don't think we do ourselves any favors by following StudioPress. Not because StudioPress isn't good, but because it isn't Understrap and we're here because understrap.

I could definitely support a limited filter setup for classes, but on that journey, I wonder if there are any other ways to handle this. We're Underscores + BootStrap.

Maybe we could add a configuration file that chooses things like light/dark, container-size, breakpoints, and others with sane defaults and pulls those into the theme instead of individual filters to change throughout.

That's just something I thought of while reading through this thread, and maybe it's an even heavier lift, but it's more Bootstrappy to me, and I think it probably focuses on 90% of what we might want to use filters for, makes them accessible, and it centralizes them.

I'm also converting this to a discussion instead of an issue.
